### DISCUSSION TITLE
Config for ruflin/elastica 2.3.3, laravel 5.8 with php 7.4 and elasticsearch 7.13

### DISCUSSION THREAD

**Discussion by ghost** [ruflin/Elastica#1999] [Ideas]
For this to work with laravel 5.8 the config is:
```
        'elastic' => [
            'driver' => 'monolog',
            'handler' => \Monolog\Handler\ElasticSearchHandler::class,
            'with' => [
                'client' => \resolve(\Elastica\Client::class, [
                    'config' => [
                        'host' => \env('ELASTIC_HOST', '127.0.0.1'),
                        'port' => \env('ELASTIC_POST', '9200'),
                        'headers' => ['Content-Type' => 'application/json']
                    ],
                    'callback' => function (\Elastica\Connection $connection, \Exception $e, \ELastica\Client $client) {
                        Log::channel('single')->error($e->getMessage() . ', FOR CONNECTION: ' .
                            \json_encode($connection->toArray()) . ', LAST REQUEST: ' . $client->getLastRequest());
                    }
                ]),
                'options' => [
                    'ignore_error' => true,
                ],
                'level' => \env('LOG_LEVEL', 'error'),
            ],
            'formatter' => \Monolog\Formatter\ElasticaFormatter::class,
            'formatter_with' => [
                'index' => \env('ELASTIC_LOGS_INDEX', 'default') . \Carbon\Carbon::now()->format('-Y.W'),
                'type'  => '_doc',
            ],
        ],
```

For php 7.4, elasticsearch 7.13 and $client->getLastRequest() to work a vendor patch is needed.

        Index: vendor/ruflin/elastica/lib/Elastica/Client.php
        IDEA additional info:
        Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
        <+>UTF-8
        ===================================================================
        diff --git a/vendor/ruflin/elastica/lib/Elastica/Client.php b/vendor/ruflin/elastica/lib/Elastica/Client.php
        --- a/vendor/ruflin/elastica/lib/Elastica/Client.php	
        +++ b/vendor/ruflin/elastica/lib/Elastica/Client.php	(date 1634637724227)
        @@ -617,7 +617,7 @@
                     $request = new Request($path, $method, $data, $query, $connection);
         
                     $this->_log($request);
        -
        +            $this->_lastRequest = $request;
                     $response = $request->send();
         
                     $this->_lastRequest = $request;
        Index: vendor/ruflin/elastica/lib/Elastica/Transport/Http.php
        IDEA additional info:
        Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
        <+>UTF-8
        ===================================================================
        diff --git a/vendor/ruflin/elastica/lib/Elastica/Transport/Http.php b/vendor/ruflin/elastica/lib/Elastica/Transport/Http.php
        --- a/vendor/ruflin/elastica/lib/Elastica/Transport/Http.php	
        +++ b/vendor/ruflin/elastica/lib/Elastica/Transport/Http.php	(date 1634637702851)
        @@ -93,7 +93,7 @@
         
                 if (!empty($headersConfig)) {
                     $headers = array();
        -            while (list($header, $headerValue) = each($headersConfig)) {
        +            foreach ($headersConfig as $header => $headerValue) {
                         array_push($headers, $header.': '.$headerValue);
                     }

You should put this into readme for ruflin/elastica: 2.3.3

**Comment by ruflin** [comment]
Do I read this correctly, that you are using Elastic 2.3.3 with Elasticsearch 7.x? This is not really compatible with each other, a newer version of Elastica should be used. I might miss something here?

**Reply by ghost** [reply]
Yes you are. It works, at least to the level I need it to. I can see the logs in elasticsearch. I tested in php 8 laravel 8 and ruflin/elastica latest version:

    'elastic' => [
            'driver' => 'monolog',
            'handler' => \Monolog\Handler\ElasticaHandler::class,
            'with' => [
                'client' => \app()->make(\Elastica\Client::class, [
                    'config' => [
                        'host' => \env('ELASTIC_HOST', '127.0.0.1'),
                        'port' => \env('ELASTIC_POST', '9200'),
                    ],
                    'callback' => function (\Elastica\Connection $connection, \Exception $e, \ELastica\Client $client) {
                        \Illuminate\Support\Facades\Log::channel('single')->error($e->getMessage() .
                            ', FOR CONNECTION: ' . \json_encode($connection->toArray()) . ', LAST REQUEST: ' .
                            $client->getLastRequest());
                    }
                ]),
                'options' => [
                    'ignore_error' => true,
                ],
                'level' => \env('LOG_LEVEL', 'error'),
            ],
            'formatter' => \Monolog\Formatter\ElasticaFormatter::class,
            'formatter_with' => [
                'index' => \env('ELASTIC_LOGS_INDEX', 'default'),
                'type'  => '_doc',
            ],
        ],

This also works, sort of, but I choose elasticsearch/elasticsearch library for laravel 8 php 8 because yours contained more escaping characters than it in the message (I don't know why).

Also both are sending datetime and I needed @timestamp field because I use it together with filebeat for backup if the elasticsearch direct logging fails.

It took me some time to reverse engineer how should I use your library in config/logging.php that is why I am writing you these configs for you to put them in readme for each version.:)

**Reply by ruflin** [reply]
Where in the docs would you expect to see this? Wouldn't this more belong into laravel docs? The nice part about having now discussions in the Github repo is that users will also find it here.

@mariusdevnest Great to see you started use the Discussion part, as you can see this is still pretty new to the Elastica project.

**Reply by ghost** [reply]
@ruflin Ok. Yes the discussions part is ok. There is something in the laravel docs, but not specific to your library.
